cdap-api/src/main/java/co/cask/cdap/api/lineage/field/JoinOperation.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.lineage.field;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Represents a join operation. Input to the join operation is represented by origins(operations).
```

Comment:
I am still not happy with this class. Something seems odd. Perhaps your previous version was better. Because each side of the join can have fields with different origins. So we would have 

- a left side List&lt;InputField>
- a right side List&lt;nputField>
- the output fields are the union of left and right field names. They don't need to be declared

The constraint says "these fields from left must be the same as those fields from the right". Since we require uniqueness of field names across left and right, that could be two lists of field names. Or simply a number K, with the convention that the join is on the first K fields of left and right side fields? I am not sure. 

I also think now (contrary to my initial thinking) that the join type might not be needed, for simplicity that can be part of the description if needed.

Reply:
Anyway, for simplicity, it seems that it would be good if we can do without inner helper classes (Input and Constraints)

Reply:
Discussed with @anew. The purpose of accepting origins for Join, Merge operations was simplicity in the API.. but if thats not satisfying what we want these operations can still be represented using `TransformOperation` which takes `List<InputField>`, so we decided to remove these operations from the current PR. We will revisit if the operations are actually needed.

Reply:
Created sub-task here to track it - https://issues.cask.co/browse/CDAP-13298